Fix unguarded availability warnings in Messaging

Split from #385

Xcode 9 introduces warnings for unguarded API calls introduced in later versions. 

Occurrences show up in Auth and Messaging:

>  $ pod spec lint FirebaseCommunity.podspec
> 
>  -> FirebaseCommunity (0.1.4)
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthAppDelegateProxy.m:122:10: warning: 'UIApplicationOpenURLOptionsAnnotationKey' is only available on iOS 9.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:535:51: note: 'UIApplicationOpenURLOptionsAnnotationKey' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthAppDelegateProxy.m:122:10: note: enclose 'UIApplicationOpenURLOptionsAnnotationKey' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:47:3: warning: 'SFSafariViewController' is partial: introduced in iOS 9.0 [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/System/Library/Frameworks/SafariServices.framework/Headers/SFSafariViewController.h:27:12: note: 'SFSafariViewController' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.h:41:12: note: annotate 'FIRAuthURLPresenter' with an availability attribute to silence
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:104:40: warning: 'SFSafariViewController' is partial: introduced in iOS 9.0 [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:104:1: note: annotate 'safariViewControllerDidFinish:' with an availability attribute to silence
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:80:10: warning: 'SFSafariViewController' is only available on iOS 9.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:80:10: note: enclose 'SFSafariViewController' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:81:33: warning: 'SFSafariViewController' is only available on iOS 9.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:81:33: note: enclose 'SFSafariViewController' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:164:3: warning: 'SFSafariViewController' is only available on iOS 9.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Auth] xcodebuild:  FirebaseCommunity/Firebase/Auth/Source/FIRAuthURLPresenter.m:164:3: note: enclose 'SFSafariViewController' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/c.cc:274:56: warning: possible misuse of comma operator here [-Wcomma]
>     - NOTE  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/c.cc:274:21: note: cast expression to void to silence warning
>     - WARN  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/c.cc:275:56: warning: possible misuse of comma operator here [-Wcomma]
>     - NOTE  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/c.cc:275:21: note: cast expression to void to silence warning
>     - WARN  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/version_set.cc:1350:5: warning: code will never be executed [-Wunreachable-code]
>     - NOTE  | [FirebaseCommunity/Database] xcodebuild:  leveldb-library/db/version_set.cc:1349:7: note: silence by adding parentheses to mark code as explicitly dead
>     - WARN  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessaging.m:407:18: warning: 'application:openURL:options:' is only available on iOS 9.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:326:1: note: 'application:openURL:options:' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessaging.m:407:18: note: enclose 'application:openURL:options:' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessaging.m:699:20: warning: 'messaging:didReceiveMessage:' is only available on iOS 10.0 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/Public/FIRMessaging.h:260:1: note: 'messaging:didReceiveMessage:' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessaging.m:699:20: note: enclose 'messaging:didReceiveMessage:' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessagingContextManagerService.m:147:20: warning: 'setAlertTitle:' is only available on iOS 8.2 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/System/Library/Frameworks/UIKit.framework/Headers/UILocalNotification.h:44:47: note: 'setAlertTitle:' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessagingContextManagerService.m:147:20: note: enclose 'setAlertTitle:' in an @available check to silence this warning
>     - WARN  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessagingRmq2PersistentStore.m:107:26: warning: 'sqlite3_errstr' is only available on iOS 8.2 or newer [-Wunguarded-availability]
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/usr/include/sqlite3.h:3385:24: note: 'sqlite3_errstr' has been explicitly marked partial here
>     - NOTE  | [FirebaseCommunity/Messaging] xcodebuild:  FirebaseCommunity/Firebase/Messaging/FIRMessagingRmq2PersistentStore.m:107:26: note: enclose 'sqlite3_errstr' in an @available check to silence this warning
> 
> Analyzed 1 podspec.

--- a/Firebase/Messaging/FIRMessaging.m
+++ b/Firebase/Messaging/FIRMessaging.m
@@ -384,7 +384,6 @@ NSString * const FIRMessagingRegistrationTokenRefreshedNotification =
   id<UIApplicationDelegate> appDelegate = application.delegate;
   SEL continueUserActivitySelector =
       @selector(application:continueUserActivity:restorationHandler:);
-  SEL openURLWithOptionsSelector = @selector(application:openURL:options:);
   SEL openURLWithSourceApplicationSelector =
       @selector(application:openURL:sourceApplication:annotation:);
   SEL handleOpenURLSelector = @selector(application:handleOpenURL:);
@@ -403,7 +402,7 @@ NSString * const FIRMessagingRegistrationTokenRefreshedNotification =
       // Do nothing, as we don't support the app calling this block
     }];
 
-  } else if ([appDelegate respondsToSelector:openURLWithOptionsSelector]) {
+  } else if (@available(iOS 9.0, *)) {
     [appDelegate application:application openURL:url options:@{}];
 
   // Similarly, |application:openURL:sourceApplication:annotation:| will also always be called, due
@@ -727,7 +726,7 @@ NSString * const FIRMessagingRegistrationTokenRefreshedNotification =
 
 - (void)receiver:(FIRMessagingReceiver *)receiver
       receivedRemoteMessage:(FIRMessagingRemoteMessage *)remoteMessage {
-  if ([self.delegate respondsToSelector:@selector(messaging:didReceiveMessage:)]) {
+  if (@available(iOS 10.0, *)) {
     [self.delegate messaging:self didReceiveMessage:remoteMessage];
   } else if ([self.delegate respondsToSelector:@selector(applicationReceivedRemoteMessage:)]) {
 #pragma clang diagnostic push

--- a/Firebase/Messaging/FIRMessagingContextManagerService.m
+++ b/Firebase/Messaging/FIRMessagingContextManagerService.m
@@ -143,7 +143,7 @@ typedef NS_ENUM(NSUInteger, FIRMessagingContextManagerMessageType) {
   }
   if ([apsDictionary[kFIRMessagingContextManagerTitleKey] length]) {
     // |alertTitle| is iOS 8.2+, so check if we can set it
-    if ([notification respondsToSelector:@selector(setAlertTitle:)]) {
+    if (@available(iOS 8.2, *)) {
       notification.alertTitle = apsDictionary[kFIRMessagingContextManagerTitleKey];
     }
   }

--- a/Firebase/Messaging/FIRMessagingRmq2PersistentStore.m
+++ b/Firebase/Messaging/FIRMessagingRmq2PersistentStore.m
@@ -104,9 +104,13 @@ typedef void(^FCMOutgoingRmqMessagesTableHandler)(int64_t rmqId, int8_t tag, NSD
 
 // Utility to create an NSString from a sqlite3 result code
 NSString * _Nonnull FIRMessagingStringFromSQLiteResult(int result) {
-  const char *errorStr = sqlite3_errstr(result);
-  NSString *errorString = [NSString stringWithFormat:@"%d - %s", result, errorStr];
-  return errorString;
+  const char *errorStr;
+  if (@available(iOS 8.2, *)) {
+    errorStr = sqlite3_errstr(result);
+  } else {
+    errorStr = "pre iOS 8.2";
+  }
+  return [NSString stringWithFormat:@"%d - %s", result, errorStr];
 }
 
 @interface FIRMessagingRmq2PersistentStore () {

